source/common/tcp_proxy/tcp_proxy.h

```diff
@@ -58,6 +58,40 @@ struct TcpProxyStats {
 class Drainer;
 class UpstreamDrainManager;
 
+/**
+ * RouteEntry is an individual resolved route entry.
+ */
+class RouteEntry {
+public:
+  virtual ~RouteEntry() = default;
+
+  /**
+   * @return const std::string& the upstream cluster that owns the route.
+   */
+  virtual const std::string& clusterName() const PURE;
+
+  /**
+   * @return MetadataMatchCriteria* the metadata that a subset load balancer should match when
+   * selecting an upstream host
+   */
+  virtual const Router::MetadataMatchCriteria* metadataMatchCriteria() const PURE;
+};
+
+/**
+ * Route holds the RouteEntry for a connection.
+ */
+class Route {
+public:
+  virtual ~Route() = default;
+
+  /**
+   * @return the route entry or nullptr if there is no route for the connection.
+   */
+  virtual const RouteEntry* routeEntry() const PURE;
```

Comment:
I think we can skip the Route/RouteEntry distinction for now and just have one Route interface.